### PR TITLE
Optimizations and cleanups

### PR DESCRIPTION
**Pull request by aaronbee**
Optimizes some hot spots seen during scale testing. The largest changes are around optimizing `toProto`.

```
name                           old time/op    new time/op    delta
MultiToProto/cellblocks-16       7.44µs ± 1%    5.40µs ± 1%  -27.34%  (p=0.000 n=9+9)
MultiToProto/no_cellblocks-16    5.22µs ± 0%    5.22µs ± 1%     ~     (p=0.423 n=10+10)

name                           old alloc/op   new alloc/op   delta
MultiToProto/cellblocks-16       6.63kB ± 0%    3.42kB ± 0%  -48.40%  (p=0.000 n=10+10)
MultiToProto/no_cellblocks-16    3.84kB ± 0%    3.84kB ± 0%     ~     (all equal)

name                           old allocs/op  new allocs/op  delta
MultiToProto/cellblocks-16         70.0 ± 0%      58.0 ± 0%  -17.14%  (p=0.000 n=10+10)
MultiToProto/no_cellblocks-16      64.0 ± 0%      64.0 ± 0%     ~     (all equal)
```

### CODE REVIEW

**Comment by aaronbee** [comment]
I still have some testing to do on a cluster to ensure these changes don't break anything.

File: region/multi.go
Comment (dethi) [inline region/multi.go:28]: TIL. I originally thought "this can't be true, Go should be smart enough to realise that slice semantic doesn't allow us to access index >= len(slice) and so it should GC them", but no :(

File: hrpc/mutate.go
Comment (dethi) [inline hrpc/mutate.go:390]: Would be nice to keep the comments for clarity

**Comment by dethi** [comment]
You will need to rebase after https://github.com/tsuna/gohbase/pull/195 is merged, that will fix the lint errors.

File: hrpc/mutate.go
Comment (aaronbee) [reply hrpc/mutate.go:390]: I had felt the code was more clear now so didn't need the comments. I'll take another look.

Comment (aaronbee) [reply hrpc/mutate.go:392]: I added a comment with the the layout of a cellblock. I didn't think the line-by-line comments of the form `copy(cellbock[i:], foo) // adding foo` were very useful. Let me know what you think.

**Comment by dethi** [comment]
Eh the data race is because of the embedded context again https://github.com/tsuna/gohbase/issues/174 😞

**Comment by aaronbee** [comment]
Bummer. I'll see what it would take to fix that.

**Comment by dethi** [comment]
If you call `m.SetContext(nil)` in `freeMulti` that would unblock here. I don't think you should try to fix https://github.com/tsuna/gohbase/issues/174 in this PR. It's a much bigger change (a welcome change but probably best to have it separated)

**Comment by aaronbee** [comment]
Oh, wow, I see the issue now. That is a bit scary. I'll leave out zeroing out the ctx field in freeMulti until #174 is resolved.

**Comment by dethi** [comment]
The change LGTM, but it would be nice to have someone else take a look as well before I merge it.

**Review by rodicus — APPROVED** [review]

File: region/client.go
Comment (rodicus) [inline region/client.go:703]: It'd be nice to have comments (similar to the commit's comments) to explain some of the magic calculations going on here

**Review by rodicus — CHANGES_REQUESTED** [review]

File: region/multi.go
Comment (aaronbee) [reply region/multi.go:35]: I could explain this better in the comments. m.calls is reused, ie. it gets appended to. By not nilling out the slice we can avoid reallocations from appends. m.regions on the other hand is just reassigned to a new slice, so there is no point in keeping the old slice around.

File: region/client.go
Comment (aaronbee) [reply region/client.go:703]: Will do.

File: region/client.go
Comment (rodicus) [inline region/client.go:752]: nit: did you mean to use fmt.Sprintf instead of fmt.Errorf?

Comment (aaronbee) [reply region/client.go:752]: It doesn't actually matter, since panic takes in an interface{} as an argument, but it's common practice to pass an error to a panic. Either way the formatted string will be printed out if it hits.

Comment (aaronbee) [reply region/client.go:752]: To actually answer your question, yes I intended to use fmt.Errorf here.

**Review by rodicus — APPROVED** [review]
LGTM

**Review by dethi — APPROVED** [review]

**Pull request merged**
